simpleviewinc/keg-hub
# Keg-CLI / Expo-CLI updates

**Ticket**: [ZEN-371](https://jira.simpleviewtools.com/browse/ZEN-371)

## Context
* I was trying to see some logs from expo to see why my SVG button was not working
  * This led me down the path of wanting the Expo dev tools
  * Which required a few updates to make work
  * A few changes are still needed, but this is a step in the right direction
  * See this [issue for more information](https://github.com/expo/expo-cli/issues/1081)
* Keep expo / expo-cli up to data
  * Keg-Hub / Taps can now use expo-cli `3.28.0`
* Fix minor bugs with keg-cli
* Add minor functionality to keg-cli

## Goal
* Make things work more-better

## Updates
* `repos/args-parse/src/argsParse.js` && `repos/args-parse/src/utils/convertNoArgs.js`
  * Added ability to set a boolean argument to false
    * Follows the standard `--no-<name of argument>`
* `repos/keg-cli/TODO.md`
  * Updated TODOs.md for keg-cli
* `repos/keg-cli/containers/tap/values.yml` && `repos/keg-cli/containers/tap/docker-compose.yml`
  * Added env for expo dev tools to work in docker container `EXPO_DEVTOOLS_LISTEN_ADDRESS`
* `repos/keg-cli/scripts/docker/docker-desktop.env`
  * Updated the docker-desktop envs to use `0.0.0.0` instead of the `127.0.0.0`
  * Fixes issues with metro bundler and expo running inside a docker container
* `repos/keg-cli/src/tasks/*/sync.js`
  * Fixed issue with calling multiple sync actions when running the sync task
  * Could only run on sync action at a time
  * During the start task, you could run multiple
  * This update allows running multiple sync actions from the sync task, matching the start task `--sync` option
* `repos/keg-cli/src/utils/services/composeService.js`
  * Updated to call the `buildContainerSync` method
* `repos/keg-cli/src/utils/services/syncService.js`
  * Export two methods 
    * `buildContainerSync` method
      * Just the  original `syncSerivce` method renamed
    * `syncSerivce` - Wrapper to run multiple sync actions when defined
      * Calls the `buildContainerSync` method for each sync action
* `repos/keg-core/package.json`
  * No longer uses the `--web-only` argument when starting expo
  * This is required for updating the expo-cli within the docker container


## Testing
**Setup**
* Check out this PR locally
  * `keg && keg pr 48`
* Run the commands
  * `keg cli env unset --key EXPO_CLI_VERSION --confirm false --comment false`
  * `keg cli env set EXPO_CLI_VERSION=3.27.14 --confirm false`
  * Or, you could
    * Open your `defaults.env` file at `~/.kegConfig/defaults.env`
    * Find the ENV `EXPO_CLI_VERSION=3.22.3`
    * Update it's value to be `3.28.0`
    * It should look like this => `EXPO_CLI_VERSION=3.28.0`
* Kill all running docker containers
  * `keg dc rm`
    * Enter the index of a container, and hit enter
    * Do this for all running containers 
  * Run `keg d clean -f` - To clean up images and containers
* Ensure no docker images or containers are exist
  * Use `keg di` && `keg dc` to see a list for each.
      * Nothing should exit

**Test 1**
* Rebuild the base image `keg base build`
  * It should now build with the latest expo-cli
* Run the `events-force` tap
  * `keg evf start --sync core`
    * Should build / start the container with the updated expo-cli
    * Syncing core is required due to the core `package.json` updates
    * You may get errors when the tap starts. This is due to the tap using version of the other repos not yet deployed to npm
       * We will fix this in the next test
   * Disconnect from the container by pressing `ctrl+c`
   * The container should continue running in the background
* Attach to the events force tap container
  * `keg evf att`
  * Once attached run `expo -V` to see the expo version being used
    * It should print out `3.28.0`
* Disconnect from the container by typing `exit`

**Test 2**
* Run the task `keg evf sync components:install:build,retheme:install:build,jsutils:install:build`
 * It should run all sync actions, not just the first one
* After the actions finish check your syncs
  * `keg sy`
  * Should print a list of each synced repo

**Test 3**
* Attach to the events force tap container again
* Run `yarn web` to start the tap
  * It should **NOT** throw any errors this time 
  * Navigate to `local.kegdev.xyz`
    * Ensure the tap works as expected
